api/src/test/java/com/datatorrent/api/DefaultOuputPortTest.java

```diff
@@ -0,0 +1,122 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package com.datatorrent.api;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ *
+ */
+public class DefaultOuputPortTest
+{
+
+  DefaultOutputPort<Object> port;
+
+  /**
+   * Same thread for setup() and emit()
+   */
+  @Test
+  public void testSameThreadForSetupAndEmit()
+  {
+    port = new DefaultOutputPort<Object>();
+    port.setup(null);
+    port.emit(null);
+    // if it comes here it passes
+  }
+
+  /**
+   * setup() not called : null thread object should not cause exception
+   */
+  @Test
+  public void testSetupNotCalledAndEmit()
+  {
+    port = new DefaultOutputPort<Object>();
+    port.emit(null);
+    // if it comes here it passes
+  }
+
+  boolean pass = false;
+
+  /**
+   * Different thread for setup() and emit()
+   * @throws InterruptedException
+   */
+  @Test
+  public void testDifferentThreadForSetupAndEmit() throws InterruptedException
+  {
+    System.clearProperty(DefaultOutputPort.DT_DISABLE_CHECK);  // do not suppress the check
+    pass = false;
+    port = new DefaultOutputPort<Object>();
+    port.setup(null);
+    Thread thread = new Thread(new Runnable()
+    {
+
+      @Override
+      public void run()
+      {
+        try {
+          port.emit(null);
+          Assert.fail("No exception thrown!");
+        } catch (IllegalStateException ise) {
+          pass = ise.getMessage().startsWith("emit() not called from operator thread!");
+        }
+      }
+    });
+    thread.start();
+    thread.join();
+    Assert.assertTrue("same thread check didn't take place!", pass);
+  }
+
+  /**
+   * Different thread for setup() and emit() but suppress check property set
+   * @throws InterruptedException
+   */
+  @Test
+  public void testDifferentThreadForSetupAndEmit_CheckSuppressed() throws InterruptedException
+  {
+    System.setProperty(DefaultOutputPort.DT_DISABLE_CHECK, "true");  // suppress the check
+    pass = false;
+    port = new DefaultOutputPort<Object>();
+    port.setup(null);
+    Thread thread = new Thread(new Runnable()
+    {
+
+      @Override
+      public void run()
+      {
+        try {
+          port.emit(null);
```

Comment:
It will be better to rely on custom Sink implementation to check for emit(). This will avoid introducing `pass` that needs to be declared volatile.